internal/worker/asynccharmdownloader/downloadworker_test.go

```diff
@@ -0,0 +1,313 @@
+	"github.com/juju/worker/v4"
+	"github.com/juju/worker/v4/workertest"
+	"go.uber.org/mock/gomock"
+	gc "gopkg.in/check.v1"
+
+	"github.com/juju/juju/core/application"
+	applicationtesting "github.com/juju/juju/core/application/testing"
+	"github.com/juju/juju/core/errors"
+	"github.com/juju/juju/core/http"
+	"github.com/juju/juju/core/logger"
+	"github.com/juju/juju/core/watcher"
+	"github.com/juju/juju/core/watcher/watchertest"
+	"github.com/juju/juju/internal/charmhub"
+	loggertesting "github.com/juju/juju/internal/logger/testing"
+	"github.com/juju/juju/internal/testing"
+)
+
+type workerSuite struct {
+	baseSuite
+
+	states         chan string
+	newAsyncWorker func() worker.Worker
+}
+
+var _ = gc.Suite(&workerSuite{})
+
+func (s *workerSuite) TestWorkerConfig(c *gc.C) {
+	defer s.setupMocks(c).Finish()
+
+	cfg := s.newConfig(c)
+	c.Assert(cfg.Validate(), jc.ErrorIsNil)
+
+	cfg = s.newConfig(c)
+	cfg.ApplicationService = nil
+	c.Assert(cfg.Validate(), jc.ErrorIs, errors.NotValid)
+
+	cfg = s.newConfig(c)
+	cfg.ModelConfigService = nil
+	c.Assert(cfg.Validate(), jc.ErrorIs, errors.NotValid)
+
+	cfg = s.newConfig(c)
+	cfg.HTTPClientGetter = nil
+	c.Assert(cfg.Validate(), jc.ErrorIs, errors.NotValid)
+
+	cfg = s.newConfig(c)
+	cfg.NewHTTPClient = nil
+	c.Assert(cfg.Validate(), jc.ErrorIs, errors.NotValid)
+
+	cfg = s.newConfig(c)
+	cfg.NewDownloader = nil
+	c.Assert(cfg.Validate(), jc.ErrorIs, errors.NotValid)
+
+	cfg = s.newConfig(c)
+	cfg.Logger = nil
+	c.Assert(cfg.Validate(), jc.ErrorIs, errors.NotValid)
+
+	cfg = s.newConfig(c)
+	cfg.Clock = nil
+	c.Assert(cfg.Validate(), jc.ErrorIs, errors.NotValid)
+}
+
+func (s *workerSuite) TestWorkerStart(c *gc.C) {
+	defer s.setupMocks(c).Finish()
+
+	changes := make(chan []string)
+
+	done := make(chan struct{})
+	s.applicationService.EXPECT().WatchApplicationsWithPendingCharms(gomock.Any()).DoAndReturn(func(ctx context.Context) (watcher.Watcher[[]string], error) {
+		close(done)
+		return watchertest.NewMockStringsWatcher(changes), nil
+	})
+
+	w := s.newWorker(c)
+	defer workertest.DirtyKill(c, w)
+
+	s.ensureStartup(c)
+
+	select {
+	case <-done:
+	case <-time.After(testing.LongWait):
+		c.Fatalf("timed out waiting for worker to finish")
+	}
+
+	workertest.CleanKill(c, w)
+}
+
+func (s *workerSuite) TestWorkerCreatesAsyncWorker(c *gc.C) {
+	defer s.setupMocks(c).Finish()
+
+	appID := applicationtesting.GenApplicationUUID(c)
+
+	changes := make(chan []string)
+
+	done := make(chan struct{})
+	s.applicationService.EXPECT().WatchApplicationsWithPendingCharms(gomock.Any()).DoAndReturn(func(ctx context.Context) (watcher.Watcher[[]string], error) {
+		return watchertest.NewMockStringsWatcher(changes), nil
+	})
+	s.httpClientGetter.EXPECT().GetHTTPClient(gomock.Any(), http.CharmhubPurpose).Return(s.httpClient, nil)
+
+	s.newAsyncWorker = func() worker.Worker {
+		close(done)
+		return workertest.NewErrorWorker(nil)
+	}
+
+	w := s.newWorker(c)
+	defer workertest.DirtyKill(c, w)
```

Comment:
If your test fails, then it's standard practice to attempt to clean up the worker.

Reply:
That will still happen with `defer CleanKill` though right?

Reply:
So the pattern should be...

```
w := startWorker()

// We don't care what the error is here, best effort clean up so 
// that other tests can run without being slowed down by workers 
// that potentially won't die.
//
// Also, we only care about the reason why the test failed,
// not the reason why the CleanKill failed to potentially stop
// because we put it in a bad state.
defer workertest.DirtyKill(c, w)

// Do some work

// Test the expectations

// You want to ensure from here on, that you **must** have a
// a clean killed worker.
workertest.CleanKill(c, w)
```

That's why I always arrange them like that, so it's easier to diagnose the test failure without the clean kill becoming a red herring in the output.

Reply:
That's fair enough, I wasn't considering the failure output